File: teagit.py


#  teagit.py 

# A simple program to create a thing that
# is almost (but not quite) entirely unlike a Git repo.
 
# If Git were a drink, and a Nutri-Matic Drink
# Synthesizer attempted to make it, it would probably
# create something like this....
  
# Author: mooseman

# Acknowledgements: A very big "thank you" to the developers of 
# Git - the DVCS that rules the universe. First written by Linus 
# Torvalds and now maintained by Junio Hamano - you guys rock!  
# Also a big "thank you" to Scott Chacon for his work on the 
# "Git Community Book" - an *outstanding* resource for learning 
# all about Git!    

# This code is released to the public domain.
# "Share and enjoy"...... ;) 

 
#  TO DO - 
#  1 - Get the most basic functionality working. 
#  That is - get a file's name, mode (permissions) and 
#  its SHA1 hash.  Save these to disk (probably using pickle). 

#  2 - Focus on using SHA1 hashes - they're the key to tracking the 
#  state of a repo and its contents over time.  

#  3 - Look at implementing a "staging area" to store "adds" before they  
#  are committed. 


import hashlib, zlib, gzip, os.path   

#  A file class 
#  Note on the "mode" - to get the mode of a file or directory, 
#  do the following - 
#  numeric_mode = os.stat('file_or_dir_name')[0]  
#  This creates a tuple with the stat details. The "mode" is the first 
#  element of the tuple.  We need to convert it to octal, as follows -
#  mymode = oct(numeric_mode & 0777)

#  A blob class.  
class blob(object): 
   def init(self): 
      self.data = {}  
      self.prev = self.blobname = None 
               
   def add(self, mydata): 
      self.name = mydata 
      self.file = open(mydata, 'rb').read() 
      self.size = str(int(os.path.getsize(mydata)))  
      self.mode = '10' + oct(os.stat(mydata)[0] & 0777) 
      if self.mode in ('100644' , '100755'): 
         self.type = 'blob' 
      else: 
         self.type = 'tree'       
      self.header = self.type + " " + self.size + "\0"          
      
      #  Calculate the SHA1 hexdigest.  
      self.stuff = self.header + self.file 
      self.sha1 = hashlib.sha1(self.stuff).hexdigest()    
      self.dirname = self.sha1[0:2]  
      self.blobname = self.sha1[2:41]  
            
      #  ( To do - calculate the SHA1s for a "tree" and a "commit". )      
      #  Save in our dict.  
      self.data.update({ self.blobname: [self.dirname, self.prev, self.name, 
          self.size, self.mode ] })  
      # Now, the just-added data will be the "prev" instance for the next data to be added.  
      self.prev = self.blobname                
      
   def display(self): 
     print self.data  
     
     
#  Test the class 
a = blob() 
a.init() 
a.add('file1.txt') 
a.display()  



#  A tree class. This has the following attributes - 
#  mode, object-type (file or tree), file (or tree) name, 
#  objectID (which is the SHA1 hash of the object).  
class tree(object): 
   def init(self): 
      self.prev = self.address = self.next = 0
      self.branch = "Main" 
      self.nodedict = {} 

   def add(self, branch, data): 
      self.branch = branch 
      self.data = data        
      self.address = self.next 
      self.next += 1 
      self.nodedict.update( {self.address: (self.branch, self.prev , self.next, self.data ) }) 
      self.prev = self.address  
            
   def display(self): 
      print self.nodedict 
      
#  Run the code 
a = tree() 
a.init() 
a.add("Main", "foo bar baz") 
a.add("Main", (12, 43, 54, 68) ) 
a.display() 



class tree2(object): 
   def init(self): 
      self.data = {}   

   def add(self, mode, name, objectid): 
      self.mode = mode 
      self.name = name 
      self.object_type = name.object_type       
      self.objectid = objectid      
   
   def display(self): 
      print self.mode, self.object_type, self.name, self.objectid 
   
#  Create a few trees 
c = tree() 
d = tree() 
e = tree() 
   
         
#  A commit object. This object is made up of trees - it shows 
#  the state of a tree as at a given point in time. 
#  It also has metadata like the name of the committer, the commit 
#  date and time, and commint comments. 
#  The key is the SHA1 of the commit (which points to a tree). 
#  Then there are the SHA1s of the trees and blobs.   
class commit(object): 
   def __init__(self): 
      self.data = {} 
      self.prev = self.curr = self.next = None 
      self.object_type = "commit" 
      
   def add(self,tree): 
      self.data.update({ tree.objectid:  tree.name} ) 
      #  Update the prev, curr and next variables. 
            
     
#  A Git tag class.  
class tag(object): 
   def init(self): 
      self.data = {} 
      
   def add(self, commit, desc): 
      self.tagID = commit.objectid 
      self.desc = desc 
      self.type = "tag" 
            
           
   
#  A "teagit object" class. This can be one of the following  
#  types - blob, tree, commit, tag. 
class teagitobject(object):
   def init(self): 
      self.data = {} 
      
   def add(self, obj): 
      self.object_id = obj.object_id 
      self.name = obj.name 
      self.obj_type = obj.type 
      self.data.update({ self.objectid: [self.name, self.obj_type] } ) 
      
                     
#  A repository object. This is a container for everything 
#  above. 
class repo(object): 
   def init(self): 
      self.data = {} 

   def add(self, path, repo, objects, refs, heads, HEAD): 
      self.path = path 
      self.repo = self.path + repo 
      self.objects = objects 
      self.refs = refs 
      self.heads = heads 
      self.HEAD = HEAD 

   # A method to display the contents of a given commit 
   def show_commit(self, commitID): 
      print self.data[commitID]   
                  
   def display(self): 
      print self.repo, self.objects, self.heads 
      
      
      
            

File: teagit.cpp

//
//  A very simple parser that is almost (but not quite) 
//  entirely unlike Git..... :)  
//  Yep, this is Teagit in C++.  
//  
//  NOTE - as of now (30th May 2009) this parser requires 
//  Boost::Spirit version 2.1. At present you will need to 
//  get that version via svn. The following command will check out 
//  Boost-trunk - that will get you this version - 
//  
//  svn co http://svn.boost.org/svn/boost/trunk boost-trunk
//
//  Written by:  Andy Elvey
// 
//  Acknowledgements: I want to acknowledge (and give a huge 
//  "Thank you!" to all of the developers of Boost::Spirit, pretty 
//  much all of whom have helped me at one time or another on the 
//  Spirit mailing-list.     
//  
//  Another very big "thank you" to the developers of Git - the DVCS 
//  that rules the universe. First written by Linus Torvalds and now 
//  maintained by Junio Hamano and a "cast of thousands". 
//  ( Well, "lots" anyway.....  :)  ) 
//    
//  This parser is released to the public domain.  
//    
//  "Share and enjoy......"  :)  
//    
//#define BOOST_SPIRIT_DEBUG  ///$$$ DEFINE THIS WHEN DEBUGGING $$$///


#include <boost/config/warning_disable.hpp>
#include <boost/spirit/include/qi.hpp>

#include <iostream>
#include <fstream>
#include <vector>
#include <string>

using namespace boost::spirit;
using namespace boost::spirit::qi;
using namespace boost::spirit::ascii;
using namespace boost::spirit::qi::labels;


///////////////////////////////////////////////////////////////////////////////
//  Our grammar
///////////////////////////////////////////////////////////////////////////////

//  NOTE - For now, I've just put three commands here, but this gives the 
//  general idea.....  :)   
// 
//  NOTE 2 - I have put the commit command first in the grammar because this 
//  helps in minimising the abiguity of the grammar.  
//  In general, I have found it to be a good idea to put the longest 
//  grammar clauses first in Spirit grammars.   

template <typename Iterator>
struct teagit_grammar : grammar<Iterator, space_type> 
{
    teagit_grammar() : teagit_grammar::base_type(expression)
	
    {
		
    expression 					
	%=  +( commands ) ;  
		
	commands 
	%= git_commit_cmd
	|  git_init_cmd 
	|  git_add_cmd ; 
	
//  The "git commit" command. 
//  Here, we give two versions of the commit command - the long and 
//  the short one. This helps to reduce ambiguity in the grammar. 
    git_commit_cmd 
	%=  long_commit_cmd | short_commit_cmd ; 
	
//  Long version of commit. 
//  This has an optional "-a" for "all".  
	long_commit_cmd 
	%=  lit("git") >> lit("commit") 
	>>  lit("-m") >> sentence; 
		
//  Sentence.  
    sentence 
	%=  lexeme[ lit('"')  
    >>  *( ( print - lit('"') )  | space) 
	>>  lit('"') ];  

//  Short version of commit.  	
	short_commit_cmd 
	%=  lit("git") >> lit("commit") ; 
			
// 	The "git init" command. 
    git_init_cmd 
	%=  lit("git") >> lit("init");  	
		
//  The "git add" command 
    git_add_cmd 
	%=  lit("git") >> lit("add") 
	 >>  lit(".") | files ;
		 					 		 	
//  Files 
    files %= +( identifier ); 
		 			
//  Identifier 	                                           	
    identifier %= lexeme[ ( alpha >> *(alnum | '_' | '.' | '-' ) ) ] ;   					   		               
								   	  							   								                                                                                                                                                                                                                                                                 
   }

      rule<Iterator, space_type> expression, commands, git_init_cmd, 
	     git_add_cmd, git_commit_cmd, long_commit_cmd, short_commit_cmd, 
		 files, identifier, sentence;  		             
							      
};



int main()  
{
    std::cout << "/////////////////////////////////////////////////////////\n\n";
    std::cout << "\t\t A toy Git-like parser...\n\n";
    std::cout << "/////////////////////////////////////////////////////////\n\n";
    std::cout << "Type a Git command  \n" ; 
    std::cout << " (e.g. git init,  git add . ,  git add foo.txt bar.html ,  \n" ;
	std::cout << " git commit,  git commit -m \"First commit\"  \n" ; 
    std::cout << " Type [q or Q] to quit\n\n" ;

    typedef std::string::const_iterator iterator_type;
    typedef teagit_grammar<iterator_type> teagit_grammar;

    teagit_grammar mygrammar; 

    std::string str;

    while (std::getline(std::cin, str))
    {
        if (str.empty() || str[0] == 'q' || str[0] == 'Q')
            break;

        std::string::const_iterator iter = str.begin();
        std::string::const_iterator end = str.end();
        bool r = phrase_parse(iter, end, mygrammar, space);

        if (r && iter == end)
        {
            std::cout << "-------------------------\n";
            std::cout << "Parsing succeeded\n";
            std::cout << "-------------------------\n";
        }
        else
        {
            std::string rest(iter, end);
            std::cout << "-------------------------\n";
            std::cout << "Parsing failed\n";
            std::cout << "stopped at: \": " << rest << "\"\n";
            std::cout << "-------------------------\n";
        }
    }

    std::cout << "Bye... :-) \n\n";
    return 0;
}


